datafusion/spark/src/function/bitwise/bit_shift.rs

```diff
@@ -0,0 +1,747 @@
+
+use std::any::Any;
+use std::sync::Arc;
+
+use arrow::array::{ArrayRef, ArrowPrimitiveType, AsArray, PrimitiveArray};
+use arrow::compute;
+use arrow::datatypes::{
+    ArrowNativeType, DataType, Int32Type, Int64Type, UInt32Type, UInt64Type,
+};
+use datafusion_common::{plan_err, Result};
+use datafusion_expr::{
+    ColumnarValue, ScalarFunctionArgs, ScalarUDFImpl, Signature, Volatility,
+};
+use datafusion_functions::utils::make_scalar_function;
+
+use crate::function::error_utils::{
+    invalid_arg_count_exec_err, unsupported_data_type_exec_err,
+};
+
+/// Performs a bitwise left shift on each element of the `value` array by the corresponding amount in the `shift` array.
+/// The shift amount is normalized to the bit width of the type, matching Spark/Java semantics for negative and large shifts.
+///
+/// # Arguments
+/// * `value` - The array of values to shift.
+/// * `shift` - The array of shift amounts (must be Int32).
+///
+/// # Returns
+/// A new array with the shifted values.
+///
+fn shift_left<T: ArrowPrimitiveType>(
+    value: &PrimitiveArray<T>,
+    shift: &PrimitiveArray<Int32Type>,
+) -> Result<PrimitiveArray<T>>
+where
+    T::Native: ArrowNativeType + std::ops::Shl<i32, Output = T::Native>,
+{
+    let bit_num = (T::Native::get_byte_width() * 8) as i32;
+    let result = compute::binary::<_, Int32Type, _, _>(
+        value,
+        shift,
+        |value: T::Native, shift: i32| {
+            let shift = ((shift % bit_num) + bit_num) % bit_num;
+            value << shift
+        },
+    )?;
+    Ok(result)
+}
+
+/// Performs a bitwise right shift on each element of the `value` array by the corresponding amount in the `shift` array.
+/// The shift amount is normalized to the bit width of the type, matching Spark/Java semantics for negative and large shifts.
+///
+/// # Arguments
+/// * `value` - The array of values to shift.
+/// * `shift` - The array of shift amounts (must be Int32).
+///
+/// # Returns
+/// A new array with the shifted values.
+///
+fn shift_right<T: ArrowPrimitiveType>(
+    value: &PrimitiveArray<T>,
+    shift: &PrimitiveArray<Int32Type>,
+) -> Result<PrimitiveArray<T>>
+where
+    T::Native: ArrowNativeType + std::ops::Shr<i32, Output = T::Native>,
+{
+    let bit_num = (T::Native::get_byte_width() * 8) as i32;
+    let result = compute::binary::<_, Int32Type, _, _>(
+        value,
+        shift,
+        |value: T::Native, shift: i32| {
+            let shift = ((shift % bit_num) + bit_num) % bit_num;
+            value >> shift
+        },
+    )?;
+    Ok(result)
+}
+
+/// Trait for performing an unsigned right shift (logical shift right).
+/// This is used to mimic Java's `>>>` operator, which does not exist in Rust.
+/// For unsigned types, this is just the normal right shift.
+/// For signed types, this casts to the unsigned type, shifts, then casts back.
+pub trait UShr<Rhs = Self> {
+    type Output;
+
+    #[must_use]
+    fn ushr(self, rhs: Rhs) -> Self::Output;
+}
```

Comment:
```suggestion
trait UShr<Rhs> {
    fn ushr(self, rhs: Rhs) -> Self;
}
```

Some simplifications

- Since Rhs is always specified, no point having default
- Output type is always same as Self in implementations so can remove that
- Don't think must_use attribute is necessary here